src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests_WithElement_Constructor.cs

```diff
@@ -1361,5 +1755,93 @@ static void Main()
         CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("42"));
     }
 
+    [Fact]
+    public void WithElement_WithLambda_ToDelegate()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public Delegate ValueFunc { get; }
+                
+                public MyList(Delegate func) : base()
+                {
+                    ValueFunc = func;
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    MyList<int> list = [with(() => 42), 1];
+                    Console.WriteLine(list.ValueFunc.DynamicInvoke());
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("42"));
+    }
+
+    [Theory]
+    [InlineData("1")]
+    [InlineData("(short)1")]
+    public void WithElement_WithLambda_InferenceWithArgAndConstructor(string returnValue)
+    {
+        var source = $$"""
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public MyList(T arg) : base()
+                {
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    Goo([with(() => {{returnValue}}), () => 2]);
```

Comment:
Consider a version where the element is the one with the `short` parameter type; it would be good to have non-constant `int` value as the return value as well, so we can see it fail to convert.

Reply:
d74e347b229